Implement `til::au16` and `til::u16a` conversion functions & make first use in `WriteConsoleAImpl`

german-one opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

Implement conversion functions that complement the existing `til::u8u16` and `til::u16u8` for the use along with other codepages than UTF-8. A state class will take care of partials at the boundaries of DBCS-encoded text.

<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References

Proposed in https://github.com/microsoft/terminal/pull/4422#discussion_r374364170

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [ ] Closes #xxx
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

These functions have the potential to supersede `ConvertToW` and `ConvertToA` by adding partials handling.
NOTE: The `astate` class supports only a subset of the possible codepages (see the remarks in the code).

In order to see how it works I updated `WriteConsoleAImpl` accordingly.
I plan to search the code base for more points to apply these functions to, and commit those updates in an upcoming PR.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

- long DBCS strings split into short chunks with an odd number of bytes, and outputted to the console using `WriteConsoleA`
- unit test added

DHowett-MSFT commented:
Initial thought: I'm not sure I love that it complicates `u8state`; I don't think that all u8-only consumers should be forced to have a codepage check and branch because conhost has some A-codepage legacy. I dunno; @miniksa?

miniksa commented:
> Initial thought: I'm not sure I love that it complicates `u8state`; I don't think that all u8-only consumers should be forced to have a codepage check and branch because conhost has some A-codepage legacy. I dunno; @miniksa Michael Niksa FTE?

I have not read the code yet, but my immediate reaction is "I'd rather have `u8state` stay `u8state` and let all legacy codepages have `astate` so we can walk away from the A theoretically in Terminal."

german-one commented:
@miniksa @DHowett-MSFT 
They share a lot of code that would be redundant otherwise, And if you look at `WriteConsoleAImpl`, see you can just call `til::au16` with any codepage id incl. UTF-8. The `astate` is always the same and partials are discarded automatically if the codepage changes.

If you still don't like it i will of course divide them into two.

miniksa commented:
> @miniksa @DHowett-MSFT
> They share a lot of code that would be redundant otherwise, And if you look at `WriteConsoleAImpl`, see you can just call `til::au16` with any codepage id incl. UTF-8. The `astate` is always the same and partials are discarded automatically if the codepage changes.
> 
> If you still don't like it i will of course divide them into two.

Nah, I'll reconsider when I get to reading this. I have to fix the tests for the project first. Sorry, I've been busy doing that instead of reading PRs.

german-one commented:
Haha, I was surprised anyway to get immediate reactions from you guys. I'm not in a hurry 🙂

DHowett-MSFT commented on src/inc/til/u8u16convert.h:
```diff
@@ -145,9 +145,13 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+#pragma warning(push)
+#pragma warning(suppress : 26482) // Suppress bounds.2 check for indexing with constant expressions.
+#pragma warning(suppress : 26446) // Suppress bounds.4 check for subscript operator.
+                                for (int idx{}; cpInfo.LeadByte[idx] != 0; idx += 2) // OK because the LeadByte array is guaranteed to end with two 0 bytes.
```
FWIW: this is what `til::at` is for :smile: if it can be used here

german-one replied:
I already tried it and failed to build. But maybe I did something wrong. I was never expecting it was so hard to work with a simple array 😅

german-one replied:
Haha. I commit one more, go to bed, and see tomorrow what happened...

german-one replied:
Hm, x86 build this time 🙃 I wait for @miniksa's review before I try to figure out why this fails. Probably I have to do more when I get feedback anyway .

german-one commented:
@DHowett-MSFT I tried to make it easier for the compiler to remove the additional branching for `u8u16` calls. But maybe this is only wishful thinking.

miniksa commented:
> Haha, I was surprised anyway to get immediate reactions from you guys. I'm not in a hurry 🙂

Yeah, I don't know what got into me. Haha. I have you on my list, but I'm trying to track down the CI test failures now as top priority. So my full review will take a bit longer. Thanks for your patience.

german-one commented:
This is annoying now. I have any kind of const and non-const types in my test code, used `til::at` as rvalue and lvalue, turned all Microsoft analysis rules on, and it passed the analysis without any warning. But eventually it fails here.
Are we really not able to avoid additional suppression of warnings if we want to access an array member?

miniksa commented:
> This is annoying now. I have any kind of const and non-const types in my test code, used `til::at` as rvalue and lvalue, turned all Microsoft analysis rules on, and it passed the analysis without any warning. But eventually it fails here.
> Are we really not able to avoid additional suppression of warnings if we want to access an array element?

We're likely to approve a few `#pragma warning(suppress:)`s in library code because it keeps us from making mistakes elsewhere.

german-one commented:
> We're likely to approve a few #pragma warning(suppress:)s in library code because it keeps us from making mistakes elsewhere.

@miniksa Actually that was not the point. I still think my approach to update `til::at` was right.
The current implementation limits it to class types that have a `size()` method. IMO it should work for most to all types that support the subscript operator, such like random-access iterators, C-arrays, and pointers. We should accept signed index types because iterators and pointers support negative offsets. And we should avoid to cast the passed index to `ptrdiff_t` or `size_t` because the expected type size might be smaller. E.g. `winrt::hstring::size_type` is always `uin32_t`.
However, I'm disenchanted because I spent a few hours with testing to avoid another rejected commit. There have been enough of them in this PR for the same reason already. And I still don't understand why it shouted out that I used pointer arithmetic at the points I called `at`.

Nevermind. This is all rather off-topic now 🙂

miniksa commented:
> > We're likely to approve a few #pragma warning(suppress:)s in library code because it keeps us from making mistakes elsewhere.
> 
> @miniksa Actually that was not the point. I still think my approach to update `til::at` was right.
> The current implementation limits it to class types that have a `size()` method. IMO it should work for most to all types that support the subscript operator, such like random-access iterators, C-arrays, and pointers. We should accept signed index types because iterators and pointers support negative offsets. And we should avoid to cast the passed index to `ptrdiff_t` or `size_t` because the expected type size might be smaller. E.g. `winrt::hstring::size_type` is always `uin32_t`.
> However, I'm disenchanted because I spent a few hours with testing to avoid another rejected commit. There have been enough of them in this PR for the same reason already. And I still don't understand why it shouted out that I used pointer arithmetic at the points I called `at`.
> 
> Nevermind. This is all rather off-topic now 🙂

It's fine. Thanks for your effort either way.

Just keep in mind that our analysis system is our best effort of what we think we need to do for static analysis. You might prove it wrong in ways that we haven't thought about yet!

german-one commented:
> You might prove it wrong in ways that we haven't thought about yet!
I can only proof my local VS wrong which didn't report any warning in the analysis. But of course it's my fault. I should have seen that additional overloads for arrays are necessary to avoid the array decay. I don't want to tinker with that any longer. At least not in this PR. Maybe I start another one if the time is right ...

german-one commented:
The list of codepages without partials handling is a little shorter now, which also means that the support is better than what we have today.

german-one commented:
Oh and CP54936 might be of interest to meet the GB 18030 standard.

german-one commented:
I have to find out what's going wrong in #4673. Probably I'll commit the bug fix here.

german-one commented:
Fixed #4673 
1) updated the code of the state class and tested manually that `printf "\xE2"; printf "\x98\xBA\n"` prints a U+263A character
2) updated the unit tests to make sure that still up to 3 partials are cached
3) updated the unit tests to make sure caching also works if the string consists of a lead byte only
4) tested manually that #4086 is still resolved

DHowett-MSFT commented:
Do you mind if I pull https://github.com/microsoft/terminal/pull/4493/commits/3003ff2ef6a08e0041a0039105de429dcd9e96a6 into a separate PR just to fix #4673?

DHowett-MSFT commented:
(Thanks for fixing this!)

german-one commented:
Yes, of course. Feel free to separate it. I would have done this if I was more familiar with git.

DHowett-MSFT commented:
Thanks for the write access! I pushed a branch to your repository called `dev/duhowett/cherry-pick-3003` that just has this one fix.

For reference, here's what I did!

```
$ git clone https://github.com/german-one/terminal.git
$ cd terminal

terminal $ git reset --hard 4a3ed3e # reset to last commit from microsoft

terminal $ git cherry-pick 3003ff2 # cherry-pick your fix

terminal $ gvim src/inc/til/u8u16convert.h # resolve the conflicts
terminal $ git add src/inc/til/u8u16convert.h
terminal $ gvim src/til/ut_til/u8u16convertTests.cpp
terminal $ git add src/til/ut_til/u8u16convertTests.cpp

terminal $ git commit --no-edit # complete the cherry-pick by committing it

# turn the current commit into a remote branch (good for doing a one-off push)
terminal $ git push HEAD:refs/heads/dev/duhowett/cherry-pick-3003
```

The `reset --hard` is destructive, so I made sure I was only doing it on a new fresh clone.

german-one commented:
So, you want me now to start a new PR from this branch, right? (sorry for my ignorance)

DHowett-MSFT commented:
Yes indeed! No need to apologize; git can be overwhelming

german-one commented:
Thank you very much for your efforts and assistance, @DHowett-MSFT !

miniksa commented:
@german-one, sorry for my delay on reviewing this, but it's going to be a bit longer. I have a lot of tasks I'm juggling right now. Apologies for the wait.

miniksa commented on src/host/screenInfo.cpp:
```diff
@@ -40,8 +40,6 @@ SCREEN_INFORMATION::SCREEN_INFORMATION(
     _textBuffer{ nullptr },
     Next{ nullptr },
-    WriteConsoleDbcsLeadByte{ 0, 0 },
-    FillOutDbcsLeadChar{ 0 },
```
Sorry, to maintain compatibility, the state really needs to be captured and stored on each individual `SCREEN_INFORMATION`. Not as a whole in the API servicing call in a static.

miniksa replied:
Also please author a test to that effect. Create a 2nd screen buffer and ensure that writing a partial into one of them isn't consumed by writing into the other one and vice versa.

german-one replied:
Oh, this is somehow surprising. How did it ever work for UTF-8 in the past? I'm already aware that the handling of other codepages is completely inconsistent compared to UTF-8. And I have on my list to investigate why this is. I didn't expect it was that worse though. It won't fit into this PR in total, However, I understand now that I already changed the behavior ...

german-one replied:
> Also please author a test to that effect.

```
// ScreenBufferTest.c
#include <windows.h>

int main(void)
{
  SetConsoleCP(CP_UTF8);
  SetConsoleOutputCP(CP_UTF8);

  DWORD written = 0;
  HANDLE hStdout = GetStdHandle(STD_OUTPUT_HANDLE);
  HANDLE hNewScreenBuffer = CreateConsoleScreenBuffer(GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, CONSOLE_TEXTMODE_BUFFER, NULL);

  WriteConsoleA(hStdout, "\xF0", 1ul, &written, NULL);
  WriteConsoleA(hNewScreenBuffer, "\xF0", 1ul, &written, NULL);
  WriteConsoleA(hNewScreenBuffer, "\x9F", 1ul, &written, NULL);
  WriteConsoleA(hStdout, "\x9F", 1ul, &written, NULL);
  WriteConsoleA(hStdout, "\x98", 1ul, &written, NULL);
  WriteConsoleA(hNewScreenBuffer, "\x91", 1ul, &written, NULL);
  WriteConsoleA(hNewScreenBuffer, "\x8D", 1ul, &written, NULL);
  WriteConsoleA(hStdout, "\x80", 1ul, &written, NULL);

  Sleep(2000ul);
  SetConsoleActiveScreenBuffer(hNewScreenBuffer);
  Sleep(2000ul);
}
```

Output of commit 11602df
![Screenshot (2)](https://user-images.githubusercontent.com/46659645/76147115-47421f80-6099-11ea-9d42-b4e0e303e8ce.png)
... after 2s in the second screen buffer ...
![Screenshot (3)](https://user-images.githubusercontent.com/46659645/76147131-6640b180-6099-11ea-989a-3ee094c693db.png)

I don't have any idea of how to make a unit test out of it.

miniksa commented on src/inc/til/u8u16convert.h:
```diff
@@ -76,32 +96,148 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+                                // If the Lead Byte indicates that the last bytes in the string is a partial UTF-8 code point then cache them:
+                                //  Use the bitmask at index `sequenceLen`. Compare the result with the operand having the same index. If they
+                                //  are not equal then the sequence has to be cached because it is a partial code point. Otherwise the
+                                //  sequence is a complete UTF-8 code point and the whole string is ready for the conversion to hstring.
```
hstring? No, please continue to use UTF-16 string as we're going to use this in places in the code that have nothing to do with winrt/hstring.

german-one replied:
Haha, no hstring in the code, only in the comment. I don't recall where I copied it. Apparently I failed to update it properly though.

miniksa commented on src/inc/til/u8u16convert.h:
```diff
@@ -76,32 +96,148 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+                    {
+                        auto foundLeadByte{ false };
+
+                        // Start at the beginning of the string (or the second bvte if a lead byte has been cached in the previous call) and scan forward,
```
```suggestion
                        // Start at the beginning of the string (or the second byte if a lead byte has been cached in the previous call) and scan forward,

```

miniksa commented on src/inc/til/u8u16convert.h:
```diff
@@ -76,32 +96,148 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+#pragma warning(suppress : 26482) // Suppress bounds.2 check for indexing with constant expressions.
+#pragma warning(suppress : 26446) // Suppress bounds.4 check for subscript operator.
+                                for (int idx{}; _cpInfo.LeadByte[idx] != 0; idx += 2) // OK because the LeadByte array is guaranteed to end with two 0 bytes.
+#pragma warning(pop)
```
This is what `til::at` is for. It suppresses those things for you.

german-one replied:
@miniksa If you scroll up you may recall that I was desparate not being able to find a way to access a C-array without suppressing warnings all the time.

About `til::at`:
> The current implementation limits it to class types that have a size() method. IMO it should work for most to all types that support the subscript operator, such like random-access iterators, C-arrays, and pointers. We should accept signed index types because iterators and pointers support negative offsets. And we should avoid to cast the passed index to `ptrdiff_t` or `size_t `because the expected type size might be smaller. E.g. `winrt::hstring::size_type` is always `uin32_t`.

I tried to update `til::at` but was fed up to fail. Do you want me to update it? I'd be happy to try again.

miniksa replied:
Sorry, I didn't see it or forgot. If at isn't good enough, I'm happy to see it improve if that isn't too much work such that others can use it to solve the same problem going forward. There is also a gsl::at, I think, which might work with signed types already...

german-one replied:
Also `gsl::at` doesn't handle arrays. I'll do my best to provide a working set of overloads for `til::at` 😀

german-one replied:
@miniksa Do you have any hint why the static analysis still fails?
The screenshot shows that at least the IntelliSense confirms that the overload for an array will be called. However, the analysis here still complains about array to pointer decay while I don't get any warning in my local tests.
![image](https://user-images.githubusercontent.com/46659645/76127735-ba4c8700-6002-11ea-8bcb-63a3e641ac55.png)

german-one replied:
Eventually I removed the overloads for arrays since the static analysis doesn't seem to understand that an array decay is not necessary if it takes the array overload in consideration.
Using a combination of `*` and `&` I'm able to get the pointer to the first element in the array without triggering a warning. The updated `til::at` function is able to handle pointers properly.
That's how I call it
`til::at(&(*_cpInfo.LeadByte), idx)`
Doesn't look nice, I know.

miniksa commented on src/inc/til/u8u16convert.h:
```diff
@@ -181,7 +317,7 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
                 }
 
-                // populate the part of the string that contains complete code points only
+                // poulate the part of the string that contains complete code points only
```
```suggestion
                // populate the part of the string that contains complete code points only

```

miniksa commented on src/inc/til/u8u16convert.h:
```diff
@@ -299,56 +470,120 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+            }
+            else // ANSI codepages
+            {
+#pragma prefast(suppress \
```
Instead of suppressing twice in quick succession.... maybe
```
#pragma warning(push)
#pragma warning(disable: __WARNING_W2A_BEST_FIT, "...")
// function body
#pragma warning(pop)
```
around the whole function?

DHowett-MSFT commented on src/inc/til/at.h:
```diff
@@ -3,19 +3,82 @@
+    // Return Value:
+    // - Element in the array at the offset given by the index parameter.
+    template<class T, class U, const size_t N>
+    constexpr auto at(const T (&array)[N], const U index) -> typename std::enable_if<std::is_integral<U>::value, decltype(array[0])>::type
```
If we replace `T (&array)[N]` and `const T (&array)[N]` with `T (&&array)[N]`, can we get this down to one specialization? i don't mind too much if not :)

german-one replied:
I'll definitely try. Thanks!

german-one replied:
@DHowett-MSFT Unfortunately this still leads to an array to pointer decay. I'll remove the array overloads and work around using the pointer overload even if that was definitely not my intention 😥 Sorry for all the commits. But I can't reproduce that locally.

german-one commented:
@miniksa I'll most likely file an issue that the console API functions need an overhaul. And I'll certainly jump on this. However, whenever I investigate the code I stumble over the same thing. It's the `read` parameter of `WriteConsoleAImpl`, which is related to the `lpNumberOfCharsWritten` parameter in the actual `WriteConsoleA` function I guess. There is a semantic difference between 'read' and 'written' and I still don't understand what the actual definition of the value in `read` is. Could be:
- the bytes consumed, which is the same number of bytes that we received
- the bytes used for the output, which is the number of bytes we received, plus the bytes taken from the cache, minus the bytes captured in the cache

But how does this go together with `lpNumberOfCharsWritten`? And what is meant here?
- the number of glyphs written
- the number of character cells occupied for the output

Can you shed some light on that?

jsoref commented:
Most of these are from master (you can look at master to see what's pending). It doesn't like `au` which is I think something you added.

jsoref commented:
I made #5124 to get master to a green check mark, any items that you see in the output in https://github.com/microsoft/terminal/pull/4493#commitcomment-38052738 which aren't in https://github.com/jsoref/terminal/commit/e80b6029b4851be7830a9a1eecee720c23532790 are things that you'll need to add to a file.

That appears to be just:
* au
* GSM

That's reasonable as both appear to be present in your PR. Going forward, the spell checker should automatically annotate your commits instead of just saying "I found stuff".

I'd personally add both items to the main `whitelist.txt` file (it would like to be sorted using `sort -u -f`, but it doesn't really care too much).

german-one commented:
Great, thanks!

jsoref commented:
(And that's merged, so now it really should just be those to.)

german-one commented:
Hmm, seems that I still should have added all annotated words? Sorry for my ignorance.

jsoref commented:
You'd need to merge master again. You're missing e80b602.

(Or you can have faith that when this is merged, it'd be merged in w/ master which would have that commit and you'd be fine.)

jsoref commented:
The bot is judging your branch on its own merits, not considering that it might be merged into master.  In the normal case, you would be branching off a green check (passing) master and thus any new items will be your own responsibility. Unfortunately your branch started off before the spell checker was merged, and between when the spell checker was written and when it was merged additional items were added. You then merged in master and got the spell checker which was still annoyed at `master` at that point in time. I then wrote and had merged into master a fix to make master green. And you added things which cleared your contribution. Merging master at this point should thus give you green. (As would being merged into master, assuming everyone is happy w/ your commit.)

Sorry for the bumpy ride. It's hard getting things merged into a fast moving project. But this should be the end of that adventure. Anyone who rebases on top of master now will be past it / as would anyone who merges to master now.

german-one commented:
Even if the misspellings are not marked to be resolved automatically by the bot (as I've seen in other PRs) that should be okay now since it obviously didn't find anything in the recent commits.

jsoref commented:
Interesting... yeah, I could have the bot go back and look for previous comments and resolve them (or at least mark them as obsolete). I'll add that to my imaginary backlog.

In testing, I've actually been manually doing that on some of my test PRs, so, yeah... that's definitely worth doing.

(Right now, if it doesn't find any errors, it just won't leave a comment, and you'll get your green check mark. -- Here, you can see the check mark in https://github.com/german-one/terminal/commit/ab4e0d797bd967cdf7cb1ebb11162a980718fa3e)

german-one commented:
Way too stale now 😀
This PR doesn't reference an issue which indicates that partials handling in a couple of codepages has never been a real problem in the past.
